syncd/syncd_pfc_watchdog.cpp

```diff
@@ -0,0 +1,260 @@
+    auto it = wd.m_counterIdsMap.find(queueVid);
+    if (it != wd.m_counterIdsMap.end())
+    {
+        (*it).second->portCounterIds = counterIds;
+        return;
+    }
+
+    auto pfcCounterIds = std::shared_ptr<PfcCounterIds>(new PfcCounterIds(queueId,
+                                                                          portId,
+                                                                          counterIds,
+                                                                          {}));
+    wd.m_counterIdsMap.emplace(queueVid, pfcCounterIds);
+
+    // Start watchdog thread in case it was not running due to empty counter IDs map
+    wd.startWatchdogThread();
+}
+
+void PfcWatchdog::setQueueCounterList(
+        _In_ sai_object_id_t queueVid,
+        _In_ sai_object_id_t queueId,
+        _In_ const std::vector<sai_queue_stat_t> &counterIds)
+{
+    SWSS_LOG_ENTER();
+
+    PfcWatchdog &wd = getInstance();
+
+    sai_object_id_t portId = queueIdToPortId(queueId);
+    if (portId == SAI_NULL_OBJECT_ID)
+    {
+        return;
+    }
+
+    auto it = wd.m_counterIdsMap.find(queueVid);
+    if (it != wd.m_counterIdsMap.end())
+    {
+        (*it).second->queueCounterIds = counterIds;
+        return;
+    }
+
+    auto pfcCounterIds = std::shared_ptr<PfcCounterIds>(new PfcCounterIds(queueId,
+                                                                          portId,
+                                                                          {},
+                                                                          counterIds));
+    wd.m_counterIdsMap.emplace(queueVid, pfcCounterIds);
+
+    // Start watchdog thread in case it was not running due to empty counter IDs map
+    wd.startWatchdogThread();
+}
+
+void PfcWatchdog::removeQueue(
+        _In_ sai_object_id_t queueVid)
+{
+    SWSS_LOG_ENTER();
+
+    PfcWatchdog &wd = getInstance();
+
+    auto it = wd.m_counterIdsMap.find(queueVid);
+    if (it == wd.m_counterIdsMap.end())
+    {
+        SWSS_LOG_ERROR("Trying to remove nonexisting queue counter Ids 0x%lx", queueVid);
+        return;
+    }
+
+    wd.m_counterIdsMap.erase(it);
+
+    // Stop watchdog thread if counter IDs map is empty
+    if (wd.m_counterIdsMap.empty())
+    {
+        wd.endWatchdogThread();
+    }
+}
+
+PfcWatchdog::~PfcWatchdog(void)
+{
+    endWatchdogThread();
+}
+
+PfcWatchdog::PfcWatchdog(void)
+{
+}
+
+PfcWatchdog& PfcWatchdog::getInstance(void)
+{
+    static PfcWatchdog wd;
+
+    return wd;
+}
+
+sai_object_id_t PfcWatchdog::queueIdToPortId(
+        _In_ sai_object_id_t queueId)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr =
+    {
+        .id = SAI_QUEUE_ATTR_PORT,
+        .value =
+        {
+            .oid = queueId,
+        }
+    };
+
+    sai_status_t status = sai_metadata_sai_queue_api->get_queue_attribute(queueId, 1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to get port Id of queue 0x%lx: %d", queueId, status);
```

Comment:
user sai serialize oid/status to get strings representation which will be more meaningful

Reply:
syncd PFC WD will have a few more changes in the next iteration. I will make sure to refactor all error messages to pint serialized values in next PR.

Reply:
still not fixed